src/Engine/StateTrackers/CountTracker.java

```diff
@@ -0,0 +1,41 @@
+package Engine.StateTrackers;
+
+import java.io.Serializable;
+import java.util.HashMap;
+
+public class CountTracker<RootKeyType, InstanceKeyType> implements Serializable
```

Comment:
The name and location makes it sound like this is a) a `StateTracker` in its own right, and b) that it is doing any kind of counting on its own, when it isn't.
Not sure what a better name would be though: maybe `KeyedSpreadsheet` or something. Maybe change the template type names to `XKey` and `YKey` too since that's basically what they are, and the current names "Root" and "Instance" are--well, they don't help me to understand how this thing is structured, at any rate.

Adding some javadocs wouldn't hurt either.

Reply:
Hmm. Would you prefer `CountManager`? `CountTrackerImpl`? Maybe `CountList` or even `CountListList`?
Agreed that the name isn't the best.

`KeyedSpreadsheet` doesn't indicate anything at all, to me.
`XKey` and `YKey` are similarly unintuitive - I didn't conceive of this as mappable to Cartesian space until I read those words just now.
"Root" key is the key for a particular list of counts; I guess it's comparable to a customer at a restaurant.
"Instance" key is the key to a particular number *in* a list of counts; I guess it's comparable to the individual dishes the restaurant customer has ordered at least one of.

Yeah, docs wouldn't hurt.

Reply:
I mean, it's effectively a 2D array, just with objects instead of numeric indices. But `2DWhatever` isn't a valid variable name. `CountManager` or whatever is probably fine.